### ISSUE TITLE
Can we replace event.default() with something less magic?

### ISSUE THREAD

**Issue by jakearchibald** [slightlyoff/ServiceWorker#607]
```js
self.addEventListener(event => {
  event.respondWith(fetch(event.request));
});
```

The above is close to default browser behaviour (from the document's perspective), with the exception that `fetch()` processes redirects, so as far the document is concerned, it gets a response for the original request url. This will have an impact on resources that treat their final request url as the base url, such as pages, workers & CSS.

As a solution, we invented `event.default()`:

```js
self.addEventListener(event => {
  event.respondWith(event.default());
});
```

`event.default()` resolves with the final response, but the document is knows about the changes request goes through to make it, so it knows the final request url.

This is problematic because:

* The current spec requires the original fetch to call SW, then `event.default()` drops back into the original fetch, but then goes back to the SW to resolve `event.default()`, then back down to fetch with the actual final resolving value (see https://www.w3.org/Bugs/Public/show_bug.cgi?id=27524)
* The resolved value from `event.default()` isn't particularly useful for caching purposes. You have no access to the final request url, so you can only cache it against the original request url, so you're going to break base urls if you later use it from the cache

Note: Breaking base urls isn't always bad. If you're serving a page shell you want the base url to be the original request url, not the url of the page shell.

Note: Exposing redirects is a security no-go, as some sites use them to pass tokens around. Could this be replaced by a flag on the response indicating that fetch should update the request url with the response url?

```js
event.respondWith(
  fetch(event.request).then(response => {
    response.redirectToFinalURL = true;
    return response;
  })
);
```

**Comment by annevk** [comment]
I favor this solution over having the "magic" `default()` method, but perhaps there is another reason `default()` exists that we are missing? (Which perhaps suggests we should add another primitive.)

**Comment by slightlyoff** [comment]
I don't recall other reasons to have default except for shortcutting to the network (preventing other handlers from taking the event).

**Comment by jungkees** [comment]
https://github.com/slightlyoff/ServiceWorker/issues/235 is the ticket we followed up on, and according to those comments it seems fetch now does all the things `default()` was supposed to solve: (1) don't follow redirect for client requests (2) cast a 30[12378] redirect response to an opaque response. Couldn't find any other reasons.

@annevk fetch does enforce (1), right? Can you point me to the steps?

> a flag on the response indicating that fetch should update the request url with the response url

This sounds good.

**Comment by annevk** [comment]
@jungkees not all client requests are equal, so no. That is up to the specification defining that particular fetch.

**Comment by jungkees** [comment]
Alright. So the _manual redirect flag_ is set by those particular fetch invokers already. Then `e.default()` is not different from `fetch(e.request)` really.

**Comment by annevk** [comment]
By some (only navigation as far as I know). E.g. a fetch for a worker or shared worker can follow redirects just fine. I'd assume we follow them for service workers too (although not strictly a client in the same sense).

**Comment by annevk** [comment]
Tentative name: `redirectToURL`.

**Comment by annevk** [comment]
So I tried adding this, but we need to scope the mutability of the internal slot (or when the setter functions, either way). The most logical point seems to be when you pass a `Response` object to `respondWith()` or when the promise you pass to `respondWith()` has its IDL-registered set of callbacks run. Should that operation set some kind of internal flag on the `Response` object or do we expect the `Response` object to be cloned at that point in a JavaScript-only world? @heycam @bzbarsky @domenic any idea how to best explain that specification-wise?

**Comment by annevk** [comment]
So the other thing `default()` gives that I forgot about is that it preserves these bits from the request (now maybe this does not exactly work given the other issues that have come up with `default()` already):

* client
* context
* context frame type
* referrer

Dropping `default()` therefore will affect CSP, though the exact CSP story is still unclear to me. @mikewest?

**Comment by annevk** [comment]
In https://lists.w3.org/Archives/Public/public-webappsec/2015Jan/0260.html I attempted to explain the CSP "issue" to the WebAppSec WG. I still think `default()` should go.

@domenic suggests for the earlier issue that we allow mutating of `Response`'s internal `redirectToURL` slot at all times. We just need to carefully explain when the copy happens after which point it is obviously no longer observed on the original object.

**Comment by annevk** [comment]
It does seem that we should not allow modifying it for synthetic `Response` objects as they have no url to begin with. I think I'll make that throw so we can change it later once we find a suitable design.

**Comment by wanderview** [comment]
Gecko bug to implement finalURL:  https://bugzilla.mozilla.org/show_bug.cgi?id=1126815

**Comment by annevk** [comment]
Ping @mikewest @jakearchibald @jungkees. What are we going to do here?

**Comment by mikewest** [comment]
My vague solution was to have a Document check CSP both on the outgoing request (before SW intercepts), and on the incoming response (after SW manipulation). To do so, we'd need to know the final URL (or at least the final origin) in order to ensure that the Document-level CSP can actually lock `script-src` down to a certain set of origins.

**Comment by annevk** [comment]
Yeah, that is still totally doable and the plan. The main thing this changes is that service workers can no longer continue a fetch from the document while also intercepting it. (Instead they'll make a new fetch following policies of the service worker.) That was a rather weird feature to begin with that we didn't fully explain so I don't think it's a huge loss.

So at this point the next step would be for the SW specification to drop `default()`.

**Comment by wanderview** [comment]
> My vague solution was to have a Document check CSP both on the outgoing request (before SW intercepts), and on the incoming response (after SW manipulation). To do so, we'd need to know the final URL (or at least the final origin) in order to ensure that the Document-level CSP can actually lock script-src down to a certain set of origins.

Does this mean the SW script must set a URL with the correct origin when synthetically creating Response objects to return?

**Comment by annevk** [comment]
@wanderview no, synthetic responses will not have a URL and that is fine. That is how we can distinguish them from other responses and that might have some impact on CSP as well. E.g. CSP might want the option to forbid synthetic responses in certain contexts and such.

**Comment by jakearchibald** [comment]
I was trying to get my head around this, and I got in a mess because I thought `finalURL` was a string not a boolean. I don't think I'll be the last.

How about `.useFinalURL` (I was hoping to come up with multiple choices but that's all I've got).

We're going to retain that value in the cache right?

**Comment by annevk** [comment]
Yeah, I should expect all internal bits of the response to be stored, just as they need to be transfered.

**Comment by annevk** [comment]
Actually reflecting on it I don't like `useFinalURL` since there is no "final url" on responses. Maybe just `useURL`? (Indicates whether or not the `url` field is ignored.)

**Comment by wanderview** [comment]
Bikeshedding.... `urlIsFinal`?

**Comment by jakearchibald** [comment]
useURL feels like I'd give it a url to use. useOwnURL? useResponseURL?

**Comment by annevk** [comment]
What if we require an explicit `Response.redirect()` instead? Did we consider that?

**Comment by annevk** [comment]
New proposal: `respondWith(res, {overrideRequestURL: true})`. This decouples this property from `Response` which is nice as `Response` is used for other things as well. The easiest way to define it specification-wise still seems for the SW specification to set the "final url flag", but at least API-wise it is much cleaner.

**Comment by annevk** [comment]
I removed `useFinalURL`. SW should now add `overrideRequestURL` and make that set the "final url flag". I think we're all good then. The justification for having that over `Response.redirect()` is performance (less roundtrips). The justification for having that over `useFinalURL` is that it will not affect other APIs that use responses (such as Cache) and will therefore not require storing it on disk.

**Comment by jakearchibald** [comment]
Question from implementers: if the response forms a new client, but the `response.url` is out of scope, should the client change controller?

I'm leaving towards "no", the ServiceWorker controlled the page load, it should control subresources. If it wants to do something different, it could respond with a redirect (although that would double up on the final request).

**Comment by annevk** [comment]
I agree. Though that again makes me think that maybe we should wait with `overrideRequestURL` and "final url flag" until we get requests for it.

**Comment by kinu** [comment]
Adding `overrideRequestURL` support is tracked at: https://crbug.com/478573

**Comment by annevk** [comment]
Basically @jakearchibald needs to figure out whether this is actually a requirement. Then we need to figure out whether we are okay with introducing what is in effect yet a new primitive that didn't previously exist.

**Comment by jakearchibald** [comment]
I don't think this is a V1 feature. Let's wait until we have a customer for it with a strong use-case.

**Comment by jakearchibald** [comment]
`overrideRequestURL` has flaws when it comes to navigation redirects that cross scopes/origins - https://github.com/whatwg/fetch/issues/66 may be a better solution for client requests

**Comment by bhritchie** [comment]
I found this discussion after being surprised by the treatment of redirects while implementing caching with service workers on our application. Since you're looking for use-cases, I'll describe ours.

We have Rails app housing a substantial single-page javascript app (SPA). I'm caching a bunch of assets for the SPA, and then my default caching strategy is to intercept fetch events and responding with a cached asset if available, falling back to network where necessary (i.e. `return cachResponse || fetch(event.request)`). However, the fact that any redirects in the fetch are not seen by the client results in some problems for log in and log out in particular, and I expect to come across more cases as I continue testing.

Authentication for our app is handled by the popular [Devise](https://github.com/plataformatec/devise) gem. A user who logs in through our home splash page will hit `/users/sign_in` and then be redirected to /app (the actual location of the SPA). But with the caching strategy described above, users will get to the app but see the address `/users/sign_in`. Now we do routing inside the SPA as well - for example you might wind up at `/app#project/5/task/10`. Now that will instead be `/users/sign_in#project/5/task/10`. Which is no longer a meaningful url that can be shared, even if it doesn't otherwise break the app.

I can get around this by being more selective about what I intercept - for example I can whitelist certain routes or only intercept fetches for e.g. `.js` and `.css` assets or whatever. But besides the fact that this making the caching logic a little more complicated, there are routes with redirects that I *do* want to handle. For example, when a user hits `/users/sign_out` (which redirects to `/`), I dump some stuff from the cache so that the app is not still available offline even after they've logged out. But this means that `/` appears as `/users/sign_out`. (And I do have fetch `users/sign_out` from the service worker because I need to make sure I'm waiting on the cache cleanup or worse things will happen.) And again, this a big app and I expect to run into more surprises like this in less obvious places.

So what do I want? When I started with `cacheResult || fetch(event.request)`, what I really wanted or expected was something like `cacheResult || event.default()` where `event.default()` means "do whatever the browser would have done otherwise", where that includes making redirects transparent to the client. That would work for the sign in case, though not the sign out case. So something like a `finalURL` or `overrideRequestURL` option sounds good too, for routes that do require some kind of special handling. I have also seen suggestions about allowing service workers to control client navigation - that sounds like a neat feature, but it also sounds like more responsibility than I want for this kind of case. I have an app that already works well, and as far as possible I want it to just keep behaving as it does already.

Thanks for all your work on this regardless!

**Comment by jakearchibald** [comment]
Thanks for the detailed use-case. In the mean time, you can check `response.url`, and if it doesn't match `request.url` you can respond with a redirect. However, this will result in a double request for the final url which isn't great.

You're not the first to have this problem, my current thinking is at https://github.com/whatwg/fetch/issues/66, although we've got to make sure we're not breaking security.

**Comment by annevk** [comment]
@bhritchie you could also do
```js
if(cacheResponse)
  event.respondWith(cacheResponse)
```
and let the default handling of the service worker take over otherwise. That should do what you want.

**Comment by bhritchie** [comment]
@annevk It looks like if I wait to check the cache before calling `event.respondWith()` then it's too late:

`Uncaught (in promise) DOMException: Failed to execute 'respondWith' on 'FetchEvent': The fetch event has already been responded to.`

**Comment by annevk** [comment]
That makes sense. I guess we'd need something like `event.default()` then, but not with the same design as we had initially (it cannot return anything to the service worker).

**Comment by jakearchibald** [comment]
> Can we replace event.default() with something less magic?

Yes we can & did
